### ISSUE TITLE
Support excluding Casks from HOMEBREW_ARTIFACT_DOMAIN coverage

### ISSUE THREAD

**Issue by byjrack** [Homebrew/brew#13226]
### Provide a detailed description of the proposed feature

HOMEBREW_ARTIFACT_DOMAIN allows the use of a forward proxy that supports URL prefixing to handle all payload download functionality for brew client as noted in the [manual Envrionment section](https://docs.brew.sh/Manpage#environment).  

Some organizations use a cache mirror (e.g. JFrog Artifactory Remote) of the Packages attached to homebrew-core instead of a simple forward proxy by restricting to a URL.  This works without issue for Formulae/Bottles that are stored in GitHub Packages as the single source domain is switched out for the envvar content at download time.  Casks though will source the install payload from arbitrary internet sources as defined by the Cask author.  Since the source is not a single domain brew cannot just replace the domain with the ARTIFACT_DOMAIN content and instead prefixes the destination.  This will work if the ARTIFACT_DOMAIN target is a simple forward proxy that supports URL prefixing, but for a mirror it may fail if that mirror is restricted to a specific scope and protocol.

Adding an additional HOMEBREW_ARTIFACT_DOMAIN_NO_CASK envvar boolean indicating the flow its use is adjusting and follows other "opt out" flags in the environment variables.

https://github.com/orgs/Homebrew/discussions/3032 
https://github.com/Homebrew/brew/issues/13222 
https://github.com/Homebrew/brew/issues/12568 

### What is the motivation for the feature?

Desire to use HOMEBREW_ARTIFACT_DOMAIN to direct users to an internal GitHub Packages cache for homebrew-core which supports our mirror cache solution (JFrog Artifactory) while not influencing the download flow for Cask payloads as that requires supporting arbitrary internet destinations which a cache may not be able to support.

Currently users have to do install/updates using the envvar and then unset the var to do an install/update of Casks.

### How will the feature be relevant to at least 90% of Homebrew users?

So there is a significant number of Homebrew users in organizations that require some registry controls and security checks for applications that are being used in the firm.  I am not 100% sure it would be 90% of users, but it's a reasonable size community.  So gating brew through a system like Artifactory adds some insight, but leaving the arbitrary destinations to standard User Access proxies that are managed by other teams.

### What alternatives to the feature have been considered?

Attempts have been made to leverage Git repo cloning (to localize the Formulae) combined with HOMEBREW_BOTTLE_DOMAIN, but with limited success using them effectively together.  And BOTTLE seemed to be up in the air on if it would be deprecated so wasn't sure if it would be around in the future.

**Comment by MikeMcQuaid** [comment]
> Adding an additional HOMEBREW_ARTIFACT_DOMAIN_NO_CASK envvar boolean indicating the flow its use is adjusting and follows other "opt out" flags in the environment variables.

I wonder if this is even needed; I'm not sure anyone is successfully using `HOMEBREW_ARTIFACT_DOMAIN` with casks?

Thoughts?

**Comment by byjrack** [comment]
I know I would like to use `HOMEBREW_ARTIFACT_DOMAIN` to steer folks from sourcing everything over the Internet and the user access proxies by leveraging a cache. The change is because if you use bottles and casks at the same time you may have poor results during install/update when that is set to something other than a forward proxy.  JFrog's expectations was that the functionality just worked, but they didn't actually test it against Casks. While they can handle mirroring the GitHub Packages source they can't handle the arbitrary sourcing needed for Cask support which is why they added the comments to their KB noting the conflict.

Originally I had poked at `HOMEBREW_BOTTLE_DOMAIN` to try and scope it only to bottles but it seems to try the tar.gz structure instead of the blob/sha256 we would have in the OCI registry mirror as noted in #11781.

And if anyone has a method to allow for steering brew to a Packages mirror for bottles and falling back to standard proxy environment variables for casks I am game.  The current workaround for folks is to do their install upgrade using the mirror and then unset `HOMEBREW_ARTIFACT_DOMAIN` so that the casks that fail downloading will successfully download.

**Comment by byjrack** [comment]
And @MikeMcQuaid I was making the assumption that folks were using that existing model and was thinking of do no harm.  Now I am curious on the perimeter solution that works in that way (reverse proxy/webapp), but that's just my experience w proxy environments.  And also it is based off the docs as well where any URL gets a prefix which given for ghcr.io you actually will replace and not prefix.

Now if `HOMEBREW_ARTIFACT_DOMAIN` should only be replaced on matches for the GitHub Packages endpoint I am wondering if there is a bug in the regex.  Now I have not spent a ton of time on ruby code so I may be reading this wrong.  

Currently the url.sub(...) call that is used into when artifact_domain uses a regex with a submatch on Packages:URL which means it also matches "https" as well and will replace that with the content of artifact_domain.  If you remove that submatch on Packages (which I think was intended) than the URL substitution will only happen for ghcr.io endpoints and ignore any others (e.g. Casks).

Example
```
url1 = "https://ghcr.io/v2/homebrew/core/gnu-getopt/manifests/2.38"
url2 = "https://corretto.aws/downloads/resources/11.0.15.9.1/amazon-corretto-11.0.15.9.1-macosx-x64.pkg"
artifact_domain = "https://pkgregistry.com/Homebrew"
gpregex1 = %r{^(https?://ghcr.io/)?}o
gpregex2 = %r{^https?://ghcr.io/?}o
replaced = "#{ad.chomp("/")}/"

puts url1.sub(gpregex1,ad)
puts url2.sub(gpregex1,ad)
puts url1.sub(gpregex2,ad)
puts url2.sub(gpregex2,ad)

#Which results in 
https://pkgregistry.com/Homebrewv2/homebrew/core/gnu-getopt/manifests/2.38
https://pkgregistry.com/Homebrewhttps://corretto.aws/downloads/resources/11.0.15.9.1/amazon-corretto-11.0.15.9.1-macosx-x64.pkg
https://pkgregistry.com/Homebrewv2/homebrew/core/gnu-getopt/manifests/2.38
https://corretto.aws/downloads/resources/11.0.15.9.1/amazon-corretto-11.0.15.9.1-macosx-x64.pkg
```

**Comment by byjrack** [comment]
And I did manually patch my local installed copy and bottles coming from our package registry and following the proxy envvars out to the Internet for the casks.

So happy to open the PR for the description in env_config and the removal of the sub-match in download_strategy if you think this was the intent of  https://github.com/Homebrew/brew/pull/11766 and the regex match just got by the tests.

**Comment by MikeMcQuaid** [comment]
> And @MikeMcQuaid I was making the assumption that folks were using that existing model and was thinking of do no harm.

We can revert if necessary and people complain. I don't see a lot of evidence that any proxy supports casks in the way this variable requires.



> So happy to open the PR for the description in env_config and the removal of the sub-match in download_strategy if you think this was the intent of #11766 and the regex match just got by the tests.

Happy to review, not sure I can comment sufficiently without seeing some code. Feel free to not e.g. fix style or tests first if it makes that quicker/easier in case it gets rejected.

**Comment by byjrack** [comment]
We have a few internal corporate hurdles and I am in the queue, but I am hoping that the removal of a couple parens (and change to env_config for the docs) will cruise through the gates w/o too much back and forth.  Fortunately no CLA on this one which gets extra fun.

**Comment by byjrack** [comment]
Resolved by #13313
